### PR TITLE
backupccl: load previous incremental manifests in parallel

### LINKED ISSUE #47160: backupccl: small incremental backups slow down over time
Incremental backups that are built on top of many layers of previous incremental backups may be bottlenecked on loading all of the previous incremental backup manifests.

### PR DESCRIPTION
**Pull request by pbardea**
Previously, incremental backups would fetch the manifests of the backups
it is building off of serially. This would be a bottle-neck for small
backups. This commit issues these requests in parallel, aiming to reduce
the duration of incremental backups.

About 70 incremental backups were run in each case, each building off the last.

Before the change, we see that incremental backups take longer and longer to run (even though each incremental backup isn't backing up any data).
Screen Shot 2020-04-07 at 7 08 27 PM

After the change:
Screen Shot 2020-04-07 at 7 00 48 PM

Fixes #47160.

Release note (performance improvement): reduce time needed to run a
backup command when it is built on a lot of previous incremental
backups.

### CODE REVIEW

File: pkg/ccl/backupccl/backup_planning.go [hunk 1]
```diff
@@ -410,21 +413,30 @@ func backupPlanHook(
...
+						ctx, uriFile, makeCloudStorage, encryption,
+					)
+					if err != nil {
+						return errors.Wrapf(err, "failed to read backup from %q", uri)
+					}
+					prevBackups[j] = desc
+					return nil
+				})
 			}
+			wg.Wait()
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:439]: g.Wait()

File: pkg/ccl/backupccl/backup_planning.go [hunk 2]
```diff
@@ -410,21 +413,30 @@ func backupPlanHook(
 				}
 			}
 			prevBackups = make([]BackupManifest, len(incrementalFrom))
+			var wg sync.WaitGroup
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:416]: I don’t think we need the waitgroup — the ctxgroup has a .Wait()

File: pkg/ccl/backupccl/backup_planning.go [hunk 3]
```diff
@@ -410,21 +413,30 @@ func backupPlanHook(
...
-					ctx, uri, makeCloudStorage, encryption,
-				)
-				if err != nil {
-					return errors.Wrapf(err, "failed to read backup from %q", uri)
-				}
-				prevBackups[i] = desc
+				j := i
+				uriFile := uri
+				g.GoCtx(func(ctx context.Context) error {
+					defer wg.Done()
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:422]: Ditto: don’t need this — the ctxgroup keeps track of the count on its own when you return

File: pkg/ccl/backupccl/backup_planning.go (hunk 1 above)
Comment (pbardea) [reply pkg/ccl/backupccl/backup_planning.go]: Ah, yep. Done.

File: pkg/ccl/backupccl/backup_planning.go [hunk 4]
```diff
@@ -411,19 +413,27 @@ func backupPlanHook(
 			}
 			prevBackups = make([]BackupManifest, len(incrementalFrom))
 			for i, uri := range incrementalFrom {
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:415]: _tiny_ nit: I usually just use the index version of `range` in these types of cases, so I only have to copy that one var for the closing over thing (and then inside the fn I just do `slice[j]`). You can also shadow `i := i` to be sure no one accidentally closes over the loop var again.

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/backup_planning.go (hunk 4 above)
Comment (pbardea) [reply pkg/ccl/backupccl/backup_planning.go]: Nice - thanks for the tip! I especially like shadowing `i` in this case.

File: pkg/ccl/backupccl/backup_planning.go [hunk 5]
```diff
@@ -445,25 +455,32 @@ func backupPlanHook(
...
 					return errors.Errorf("cannot append a backup of specific tables or databases to a full-cluster backup")
 				}
 
-				for _, inc := range prev {
-					m, err := readBackupManifest(ctx, defaultStore, inc, encryption)
-					if err != nil {
-						return errors.Wrapf(err, "loading prior backup part manifest %q", inc)
-					}
-					prevBackups = append(prevBackups, m)
+				for i, inc := range prev {
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:471]: ditto here -- either shadow `inc` too or, imo, better yet don't bind it.

Comment (pbardea) [reply pkg/ccl/backupccl/backup_planning.go]: Ah, updated the usage in readBackupManifest but missed the other usage. I think my eyes are at their limit for the day.

**Review by dt — APPROVED** [review]
Thanks for the quick fix!

**Comment by pbardea** [comment]
TFTR!
bors r=dt

**Pull request merged**